cmd/klient/main.go

```diff
@@ -0,0 +1,135 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+
+	"k8s.io/klog/v2"
+
+	"github.com/NVIDIA/knavigator/pkg/config"
+)
+
+func mainInternal() error {
+	var addr, taskCfg string
+	flag.StringVar(&addr, "addr", "", "server address")
+	flag.StringVar(&taskCfg, "tasks", "", "comma-separated list of task config files and dirs")
+	flag.Parse()
+
+	taskconfigs, err := config.NewFromPaths(taskCfg)
+	if err != nil {
+		return err
+	}
+	if len(taskconfigs) == 0 {
+		return fmt.Errorf("missing 'tasks' argument")
+	}
+
+	taskURL, err := url.JoinPath(addr, "task")
+	if err != nil {
+		return err
+	}
+	resetURL, err := url.JoinPath(addr, "reset")
+	if err != nil {
+		return err
+	}
+
+	for _, taskconfig := range taskconfigs {
+		fmt.Printf("Starting test %s\n", taskconfig.Name)
+		if err := execTaskConfig(taskURL, resetURL, taskconfig); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func execTaskConfig(taskURL, resetURL string, taskconfig *config.TaskConfig) error {
+	var err error
+	for _, task := range taskconfig.Tasks {
+		if err = execTask(taskURL, task); err != nil {
+			break
+		}
+	}
+
+	if errReset := execReset(resetURL); errReset != nil {
+		if err == nil {
+			err = errReset
+		}
+	}
+
+	return err
+}
+
+func execTask(urlPath string, task *config.Task) error {
+	jsonData, err := json.Marshal(task)
+	if err != nil {
+		return fmt.Errorf("failed to marshal to JSON: %v", err)
+	}
+
+	resp, err := http.Post(urlPath, "application/json", bytes.NewBuffer(jsonData)) // #nosec G107
+	if err != nil {
+		return fmt.Errorf("failed to create request: %v", err)
+	}
+
+	defer resp.Body.Close() //nolint:errcheck
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response: %v", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		fmt.Printf("Error response: %s: %s\n", resp.Status, body)
+		return fmt.Errorf("%s", resp.Status)
+	}
+
+	return nil
+}
+
+func execReset(urlPath string) error {
+	resp, err := http.Get(urlPath) // #nosec G107
+	if err != nil {
+		return fmt.Errorf("failed to create request: %v", err)
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response: %v", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		fmt.Printf("Error response: %s: %s\n", resp.Status, body)
+		return fmt.Errorf("%s", resp.Status)
```

Comment:
Do you think it would be worthwhile for the client to fetch the logs from the server? Are the logs from the simulator server helpful in debugging test failures or would the error be sufficient?

Reply:
maybe later?

Reply:
Okay